### PR TITLE
RMB-843: Use percent encoding, not url encoding, for HTTP requests

### PR DESCRIPTION
**Pull request by julianladisch**
RFC 3986 mandates that only percent encoding is used for
HTTP requests. Using a plus to encode a space is wrong:
https://datatracker.ietf.org/doc/html/rfc3986

Wrong encoding of status == open:
`?query=status+%3D%3D+open`

Correct encoding of status == open:
`?query=status%20%3D%3D%20open`

Approach:

Implement PercentCodec.

Put a warning into the StringUtil.urlEncode javadoc.

Replace StringUtil.urlEncode calls by PercentCodec calls.

### CODE REVIEW

**Review by hjiebsco — COMMENTED** [review]
Is there any existing url encode library we can use? If not, will it work if we just replace `+` with `%20` after the call to `URLEncode`? BTW, have we seen issues due to `+` in practice?

**Review by adamdickmeiss — COMMENTED** [review]
A lot of work no existing utility? But hopefully it's been fun. ClientGenerator.java is also broken then.. As it generates clients using URLEncoder.encode for query parameter values. Should use the new utility too.

File: util/src/main/java/org/folio/util/PercentCodec.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+    throw new UnsupportedOperationException();
+  }
+
+  private static char [][] getEncodeArray() {
+    char [][] encode = new char [256][];
+    for (char c : UNRESERVED) {
+      char [] cc = { c };
+      encode[c] = cc;
+    }
+    for (int i=0; i<256; i++) {
```
Comment (adamdickmeiss) [inline util/src/main/java/org/folio/util/PercentCodec.java:40]: This wouldn't be accepted Google Code Style. (blanks around binary ops)

File: util/src/main/java/org/folio/util/PercentCodec.java [hunk 2]
```diff
@@ -101,7 +103,7 @@ public static Appendable encode(Appendable appendable, CharSequence charSequence
    * // yields "https://example.com/users?query=status%20%3D%3D%20open"
    * </pre>
    *
-   * @param charSequence the input in UTF8 encoding; if null or empty then "" is returned
+   * @param charSequence in UTF8 encoding; if null or empty then "" is returned
```
Comment (adamdickmeiss) [inline util/src/main/java/org/folio/util/PercentCodec.java:106]: The parameter is called `in` but you are referring to it as `charSequence`.

File: util/src/main/java/org/folio/util/PercentCodec.java [hunk 3]
```diff
@@ -126,7 +128,7 @@ public static CharSequence encode(CharSequence in) {
    * <p>All characters except these get percent encoded: The letters a-z and A-Z and the digits 0-9, and the
    * {@code -._~} characters.
    *
-   * @param charSequence the input in UTF8 encoding; if null or empty then "" is returned
+   * @param charSequence in UTF8 encoding; if null or empty then "" is returned
```
Comment (adamdickmeiss) [inline util/src/main/java/org/folio/util/PercentCodec.java:131]: Same here. javadoc and parameter inconsistent.

**Review by adamdickmeiss — APPROVED** [review]

**Review by adamdickmeiss — APPROVED** [review]
Thx

**Comment by julianladisch** [comment]
https://commons.apache.org/proper/commons-codec/apidocs/org/apache/commons/codec/net/PercentCodec.html
is a library that provides PercentCodec.
But the method signature `byte[] encode(byte[] bytes)` is not really useful.
The library comes with lots of other classes that we don't need, total size 345 KB.
The library is not in common use, only two FOLIO modules use it:
* mod-users-bl uses it for urlencode only, this can be replaced by RMB PercentCodec
* data-import-processing-core uses it for Base64, this should be replaced by java.util.Base64 that Java ships with since version 1.8.

File: util/src/main/java/org/folio/util/PercentCodec.java (hunk 3 above)
Comment (julianladisch) [reply util/src/main/java/org/folio/util/PercentCodec.java:131]: Thanks, fixed.

File: util/src/main/java/org/folio/util/PercentCodec.java (hunk 2 above)
Comment (julianladisch) [reply util/src/main/java/org/folio/util/PercentCodec.java:106]: Thanks, fixed.

File: util/src/main/java/org/folio/util/PercentCodec.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    return out;
+  }
+
+  /**
+   * Return charSequence with percent encoding (RFC 3986) applied.
+   *
+   * <p>All characters except these get percent encoded: The letters a-z and A-Z and the digits 0-9, and the
+   * {@code -._~} characters.
+   *
+   * @param charSequence the input in UTF8 encoding; if null or empty then "" is returned
```
[the remark this answers is not in the document]
Comment (julianladisch) [reply util/src/main/java/org/folio/util/PercentCodec.java]: Thanks, fixed.

File: util/src/main/java/org/folio/util/PercentCodec.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+   * <p>Example usage:
+   *
+   * <pre>
+   * String baseUrl = "https://example.com/users";
+   * String query =  "status == open";
+   * String url = baseUrl + "?query=" + PercentCodec.encode(query);
+   * // yields "https://example.com/users?query=status%20%3D%3D%20open"
+   * </pre>
+   *
+   * @param charSequence the input in UTF8 encoding; if null or empty then "" is returned
```
[the remark this answers is not in the document]
Comment (julianladisch) [reply util/src/main/java/org/folio/util/PercentCodec.java]: Thanks, fixed.

**Pull request merged**